astropy/modeling/core.py

```diff
@@ -1179,6 +1250,38 @@ def prepare_inputs(self, *inputs, **kwargs):
         _validate_input_shapes(inputs, self.inputs, n_models,
                                model_set_axis, self.standard_broadcasting)
 
+        # Check that the units are correct, if applicable
+
+        if self.input_units is not None:
```

Comment:
Is this logic not very similar to that of `u.quantity_input`? If you made `input_units` return a dict it would become `**kwargs` to `quantity_input` and you could drop just use the decorator on the call to `evaluate` if `input_units` is not None?